core/server/controllers/frontend/index.js

```diff
@@ -66,11 +66,21 @@ frontendControllers = {
                 return next();
             }
 
-            // If we're ready to render the page but the last param is 'edit' then we'll send you to the edit page.
+            // CASE: we only support /:slug format for pages
+            if (post.page && post.url !== req.path) {
+                return next();
+            }
+
+            // CASE: last param is of url is /edit, redirect to admin
             if (lookup.isEditURL) {
                 return res.redirect(config.paths.subdir + '/ghost/editor/' + post.id + '/');
             }
 
+            // CASE: permalink is not valid anymore, we redirect him to the correct one
+            if (post.url !== req.path) {
+                return res.redirect(post.url);
```

Comment:
In my opinion Its fine when the browser gets 301, because the new redirected URL is "now" the new "permanent" link. So the browser can ask for the new url from now on. When the blog owner changes the TZ again. Also this new URL would get a redirect. And so on.
In my opinion 301 is the right choice, because it's not temporary.